source/common/http/utility.cc

```diff
@@ -747,5 +750,116 @@ std::string Utility::PercentEncoding::decode(absl::string_view encoded) {
   return decoded;
 }
 
+const Utility::AuthorityAttributes Utility::parseAuthority(const absl::string_view& authority,
+                                                           uint32_t default_port) {
+  Utility::AuthorityAttributes auth_attr;
+  auth_attr.port = default_port;
+  auth_attr.host = authority;
+
+  bool is_ipv4 = true;
+  bool is_ipv6 = false;
+  bool have_port_ipv4 = false;
+  bool have_port_ipv6 = false;
+
+  // This section applies first-match-wins algorithm described from
+  // RFC3986(https://tools.ietf.org/html/rfc3986#section-3.2.2). Host literal include IP-literal,
+  // IPv4 Address or reg-name. In this function, We consider that IP-literal and IPv4 address as
+  // pure IP Address and return is_ip_address as true.
+
+  // check if passed authority matches IP-literal. IP-literal is constructed from bracket symbols
+  // and IPv6 address or future version of ip address. In this case, future version of ip address
+  // should be ignored
+  if (authority.find("[") != absl::string_view::npos) {
+    is_ipv6 = true;
+  }
+
+  if (authority.rfind("]:") != absl::string_view::npos && is_ipv6) {
+    have_port_ipv6 = true;
+  }
+
+  // check if passed authority matches IPv4 Address. IPv4 address is separated four section by dot
+  // deliminator, and each section has characters which is constructed from only digits
+  int delim_count = 0;
+  for (auto&& ch : authority) {
```

Comment:
I think this loop does not correctly identify IPv4 dot notation addresses. It will identify 1.2.3.9999999 as IP address. I think we should use getaddrinfo with the AI_NUMERICHOST flag to identify IPv4 addresses. Otherwise it will be hard to come up with anything simple. For reference see https://tools.ietf.org/html/draft-main-ipaddr-text-rep-00

Reply:
1.2.3.999999 is an error case anyway, no? it will fail in parseInternetAddress later, neither is it valid dns name. The spec you mentioned is expired draft.

Reply:
Ah yes, I should have scanned it a bit more to see that inet_pton is used below and we are limiting ourselves to the decimal dot notation. Reference was not the right work for this link, I have provided it as a way to see that detecting all possible representations of IPv4 addresses is a pain. But if we limit ourselves to just decimal dot notation, which is reasonable these days, it makes it much easier.